MS SQL - take 3

aldum opened the pull request:
This PR builds on the work in #29 and #60 by @kiendang and @jtjeferreira.

What it does:
- rebase previous works on top of current `main`
- fix table name escape
- fix `DataTypesTests` and `OptionalsTests` with the point of contention being the lack of a dedicated boolean column type
- add more tests (inserting both values of boolean, updating BIT columns) and document any result deviations

How this is achieved:

T-SQL (the dialect used by MS SQL Server) does have `Boolean`s as part of filter expressions, but not as column data, with `BIT`s being used as a substitute. Hence, these uses which are both `Boolean` on the Scala side, need to be distinguished. To this end, a marker was introduced in `Context`, explicitly designating `InsertValues`, `InsertSelect`, `InsertColumns` and `Update` expression as using them as values. This was done globally, because overloading would lead to a lot more boilerplate, while simply disregarding the flags value leave the existing dialects intact.

Known omissions:
- `ExprBooleanOps` for `BIT` values: small edge case, it's possible and there's syntax for it (bitwise operators), but it needs another branching of the above kind, to not break the normal OR/AND/NOT syntax. Bitwise operators are also already tested as part of NumericOps.
- RETURNING: this is not a thing in T-SQL, however it does have OUTPUT, which could be used to achieve similar results
- ON CONFLICT: similarly, not present in T-SQL, there is MERGE in the standard, but the opinions are mixed on whether it achieves the adequate level of isolation as Postgres's ON CONFLICT does.
- LATERAL JOIN: again, very Postgres, maybe CROSS APPLY could be used to achieve similar effect.
- bare VALUES: not supported. However, it is not it's own `Dialect` trait yet, which would indicate this more cleanly,

lihaoyi commented:
@kiendang @jtjeferreira could I ask if you guys would be free to help review this? Since you've both struggled with the problem, would love to get your eyes on this PR!

aldum commented:
Let me elaborate a bit.

What this does: 
- first, fixup the previous work to be compatible with changes to `main` since
- fix the `DataTypesTests` and `OptionalsTests` cases where it had errors
- add a couple more test cases focused on bools and the dialect

How this is achieved:
Due to the weird duality of not having an actual `Boolean` type, but having `Boolean` expressions (in WHERE, etc), there needs to be a disctinction of the usage, whether it's a column value being inserted/updated, or a filter condition. To this end, I introduced a marker to distinguish the `Context` in `InsertValues`, `InsertSelect`, `InsertColumns` and `Update` expressions. (The latter one also didn't have tests to catch it, also added). Now, I did this globally, because it's less lines of code, while the reading it is only happening in the dialect.

Omissions: 
- `ExprBooleanOps` for `BIT` values: small edge case, it's possible and there's syntax for it (bitwise operators), but it needs another branching of the above kind, to not break the normal OR/AND/NOT syntax
- RETURNING: this is not a thing in T-SQL, however it does have OUTPUT, which might be used to achieve similar results
- ON CONFLICT: similar, not present in T-SQL, there is MERGE in the standard, but the opinions are mixed on whether it achieves the adequate level of isolation Postgres's ON CONFLICT does.
- LATERAL JOIN: again, very Postgres, maybe CROSS APPLY could be used to achieve similar effect.
- bare VALUES: this is not it's own Dialect trait yet, which it would need to be to indicate that MSSQL can't do it

I might add some of these later as a followup.

Quirks:
- [As noted](https://github.com/com-lihaoyi/scalasql/pull/60#issuecomment-2577915790) by @jtjeferreira , there's a difference between Scala 2 and 3, [stated here](https://github.com/com-lihaoyi/scalasql/pull/94/commits/de6b77a9e856b27022e06fe382bba032c371a43c#diff-d7eb0920a6e16d135166af7944b05b192533deca3d7812b78b830ac9ac5122b6R371), this special casing is only required for the Scala 3 version. @mrdziuban has some [very interesting insights](https://github.com/com-lihaoyi/scalasql/pull/60#issuecomment-2583550166) about it, but I only read this after the fact and my macro-fu was not strong enough to determine what the source of the difference is.

jtjeferreira reviewed (commented):
Excellent work. This is a first review from github UI. I still want to checkout this branch and play a bit more with it

lihaoyi commented:
@aldum I think this looks great, thanks for your work.

For the bounty, @kiendang has already received 500USD for his work earlier, and I'll split the remaining 1000USD bounty between @aldum and @jtjeferreira at 500USD each since all of you contributed significantly to the final PR

@aldum @jtjeferreira do email me your international bank transfer credentials and I will close out the bounty
